tests/integration/construct_component_methods/testcomponent/index.ts

```diff
@@ -0,0 +1,92 @@
+// Copyright 2016-2021, Pulumi Corporation.  All rights reserved.
+
+import * as pulumi from "@pulumi/pulumi";
+import * as provider from "@pulumi/pulumi/provider";
+
+interface ComponentArgs {
+    first: pulumi.Input<string>;
+    second: pulumi.Input<string>;
+}
+
+class Component extends pulumi.ComponentResource {
+    public readonly first!: pulumi.Output<string>;
+    public readonly second!: pulumi.Output<string>;
+
+    constructor(name: string, args: ComponentArgs, opts?: pulumi.ComponentResourceOptions) {
+        if (opts?.urn) {
+            args = <any>{
+                first: undefined,
+                second: undefined,
+            };
+        }
+
+        super("testcomponent:index:Component", name, args, opts);
+
+        // Skip further initialization when being initialized from a resource reference.
+        if (opts?.urn) {
+            return;
+        }
+
+        this.registerOutputs(args);
+    }
+
+    getMessage(name: pulumi.Input<string>): pulumi.Output<string> {
+        return pulumi.all([this.first, this.second, name]).apply(([frst, scnd, nm]) => `${frst} ${scnd}, ${nm}!`);
+    }
+}
+
+class Provider implements provider.Provider {
+    public readonly version = "0.0.1";
+
+    constructor() {
+        // Register any resources that can come back as resource references that need to be rehydrated.
+        pulumi.runtime.registerResourceModule("testcomponent", "index", {
+            version: this.version,
+            construct: (name, type, urn) => {
+                switch (type) {
+                    case "testcomponent:index:Component":
+                        return new Component(name, <any>undefined, { urn });
+                    default:
+                        throw new Error(`unknown resource type ${type}`);
+                }
+            },
+        });
+    }
+
+    async construct(name: string, type: string, inputs: pulumi.Inputs,
+              options: pulumi.ComponentResourceOptions): Promise<provider.ConstructResult> {
+        if (type != "testcomponent:index:Component") {
+            throw new Error(`unknown resource type ${type}`);
+        }
+
+        const component = new Component(name, <ComponentArgs>inputs, options);
+        return {
+            urn: component.urn,
+            state: inputs,
+        };
+    }
+
+    async call(token: string, inputs: pulumi.Inputs): Promise<provider.InvokeResult> {
+        switch (token) {
+            case "testcomponent:index:Component/getMessage":
+                const self = inputs.__self__;
+                if (!pulumi.Output.isInstance(self)) {
+                    throw new Error(`__self__ is not an instance of Output for method ${token}`);
+                }
+                return {
+                    outputs: {
+                        message: self.apply(s => (<Component>s).getMessage(inputs.name)),
```

Comment:
This is annoying. `inputs.__self__` is an `Output` here (I think because the resource reference has a dependency on itself, so in our `server.ts` provider implementation we wrap it in an `Output` rather than the plain value; need to double check, though).

If my suspicion for why this is an output is correct, I'm considering special casing this so we don't wrap `__self__` as an `Output` so you don't have to use an `apply` here.

(Though, ultimately, the long term plan is that you wouldn't have to write this code yourself and we'd automatically be able to dispatch for you).

Reply:
Hm, is this the reason why we have changes elsewhere to wait for outstanding RPC requests? This effectively forces you to create resources inside of apply.

Reply:
nvm, I see that was just a refactor to keep the monitor alive. 

This makes me nervous though. Would be great to see a non-trivial test that creates a few resources within the apply (ideally with some artificially latency introduced). I wouldn't be surprised if the current approach could occasionally result in dropped dependencies.

Reply:
I've fixed this by ensuring the resource reference remains a plain value.